test/gtest/ucp/test_ucp_proto.cc

```diff
@@ -563,4 +605,224 @@ UCS_TEST_P(test_proto_perf_add_func, intersect_middle)
+     *        |      |                 |          |
+     *        |      |                 |          |
+     *        +------+--2.local_cpu1 --+          |
+     *               |                            |
+     *               |                            |
+     *               +--3.local_cpu2 -------------+
+     *
+     */
+    perf_ptr_t perf1 = create();
+    add_func(perf1, 500, 2999, UCP_PROTO_PERF_FACTOR_LOCAL_TL, local_tl_func);
+    UCS_TEST_MESSAGE << perf1.get();
+
+    perf_ptr_t perf2 = create();
+    add_func(perf2, 1000, 1999, UCP_PROTO_PERF_FACTOR_LOCAL_CPU,
+             local_cpu_func);
+    UCS_TEST_MESSAGE << perf2.get();
+
+    perf_ptr_t perf3 = create();
+    add_func(perf3, 1200, 2999, UCP_PROTO_PERF_FACTOR_LOCAL_CPU,
+             local_cpu_func);
+    UCS_TEST_MESSAGE << perf3.get();
+
+    m_perf = aggregate({perf1, perf2, perf3});
+    UCS_TEST_MESSAGE << m_perf.get();
+
+    expect_empty_range(0, 1199);
+    expect_perf(1200, 1999,
+                {{UCP_PROTO_PERF_FACTOR_LOCAL_CPU,
+                  ucs_linear_func_add(local_cpu_func, local_cpu_func)},
+                 {UCP_PROTO_PERF_FACTOR_LOCAL_TL, local_tl_func}});
+    expect_empty_range(2000, SIZE_MAX);
+}
+
+class test_proto_perf_aggregate : public test_proto_perf {
+protected:
+    using perf_vec_t = std::vector<perf_ptr_t>;
+
+    // Test aggregation of random collection of perf structures, by sampling
+    // the result at specific points
+    void test_random_funcs(unsigned num_perfs, unsigned num_segments);
+
+private:
+    // Generate a random ascending sequence of numbers, ending with SIZE_MAX
+    static std::vector<size_t> generate_random_sequence(unsigned length);
+
+    // Generate a random perf structure
+    static perf_ptr_t generate_random_perf(unsigned max_segments);
+
+    // Collect points of sampling based on perf structure segments
+    static std::vector<size_t> get_sample_points(perf_ptr_t perf);
+
+    // Calculate the expected aggregation result at a given point. If the
+    // expected result is undefined, return nullptr.
+    std::unique_ptr<perf_factors_map_t>
+    expected_aggregate_result(const std::vector<perf_ptr_t> &perfs,
+                              size_t point);
+};
+
+std::vector<size_t>
+test_proto_perf_aggregate::generate_random_sequence(unsigned length)
+{
+    std::vector<size_t> sequence;
+    size_t value = 0;
+    std::generate_n(std::back_inserter(sequence), length, [&value]() {
+        value += ucs::rand_range(10000);
+        return value;
+    });
+    sequence.emplace_back(SIZE_MAX);
+    return sequence;
+}
+
+test_proto_perf::perf_ptr_t
+test_proto_perf_aggregate::generate_random_perf(unsigned max_segments)
+{
+    perf_ptr_t perf = create();
+    size_t start    = 0;
+    auto points     = generate_random_sequence(max_segments);
+    for (auto point : points) {
+        if (ucs::rand_range(4) > 0) {
+            add_func(perf, start, point, UCP_PROTO_PERF_FACTOR_LOCAL_TL,
```

Comment:
What do you think about adding functions to several different factors? Seems like `expected_aggregate_result` already support checking such combinations.

Reply:
sounds good